sklearn/utils/validation.py

```diff
@@ -981,3 +982,155 @@ def check_scalar(x, name, target_type, min_val=None, max_val=None):
+    >>> check_psd_eigenvalues([1, 2])      # nominal case
+    ... # doctest: +NORMALIZE_WHITESPACE
+    array([1, 2])
+    >>> check_psd_eigenvalues([5, 5j])     # significant imag part
+    ... # doctest: +ELLIPSIS +NORMALIZE_WHITESPACE
+    Traceback (most recent call last):
+        ...
+    ValueError: There are significant imaginary parts in eigenvalues (1.000000
+        of the maximum real part). The matrix is maybe not PSD, or something
+        went wrong with the eigenvalues decomposition.
+    >>> check_psd_eigenvalues([5, 5e-5j])  # insignificant imag part
+    ... # doctest: +NORMALIZE_WHITESPACE
+    array([5., 0.])
+    >>> check_psd_eigenvalues([-5, -1])    # all negative
+    ... # doctest: +ELLIPSIS +NORMALIZE_WHITESPACE
+    Traceback (most recent call last):
+        ...
+    ValueError: All eigenvalues are negative (maximum is -1.000000). The matrix
+        is maybe not PSD, or something went wrong with the eigenvalues
+        decomposition.
+    >>> check_psd_eigenvalues([5, -1])     # significant negative
+    ... # doctest: +NORMALIZE_WHITESPACE
+    array([5, 0])
+    >>> check_psd_eigenvalues([5, -5e-5])  # insignificant negative
+    ... # doctest: +NORMALIZE_WHITESPACE
+    array([5., 0.])
+    >>> check_psd_eigenvalues([5, 4e-12])  # bad conditioning (too small)
+    ... # doctest: +NORMALIZE_WHITESPACE
+    array([5., 0.])
+
+    """
+    # is the provided array in double precision (float64) ?
+    if isinstance(lambdas, np.ndarray):
+        # From https://docs.scipy.org/doc/numpy/reference/arrays.dtypes.html
+        # and https://docs.scipy.org/doc/numpy/user/basics.types.html
+        if lambdas.dtype.kind == 'f':
+            is_double_precision = (lambdas.dtype.itemsize >= 8)
+        else:
+            # default for non-float dtypes
+            is_double_precision = True
+    else:
+        # default for non-numpy inputs
+        is_double_precision = True
+
+    # note: the minimum value available is
+    #  - single-precision: np.finfo('float32').eps = 1.2e-07
+    #  - double-precision: np.finfo('float64').eps = 2.2e-16
+
+    # the various thresholds used for validation
+    # we may wish to change the value according to precision.
+    significant_imag_ratio = 1e-5
+    significant_neg_ratio = 1e-5 if is_double_precision else 5e-3
+    significant_neg_value = 1e-10 if is_double_precision else 1e-6
+    small_pos_ratio = 1e-12
+
+    # Check that there are no significant imaginary parts
+    if not np.isreal(lambdas).all():
+        max_imag_abs = abs(np.imag(lambdas)).max()
+        max_real_abs = abs(np.real(lambdas)).max()
+        if max_imag_abs > significant_imag_ratio * max_real_abs:
+            raise ValueError(
+                "There are significant imaginary parts in eigenvalues (%f "
+                "of the maximum real part). The matrix is maybe not PSD, or "
+                "something went wrong with the eigenvalues decomposition."
+                "" % (max_imag_abs / max_real_abs))
```

Comment:
```suggestion
                 % (max_imag_abs / max_real_abs))
```

Reply:
I personally avoid hanging `%` signs so I always keep the quotes in front. But if that's a style guideline in sklearn I can definitely change

Reply:
or just use `format`, but I agree that the empty quote isn't useful